model/flow/event.go

```diff
@@ -99,18 +99,28 @@ type BlockEvents struct {
 
 type EventsList []Event
 
-// EventsListHash calculates a hash of events list,
-// by simply hashing byte representation of events in the lists
-func EventsListHash(el EventsList) (Identifier, error) {
-
-	hasher := hash.NewSHA3_256()
+// EventsListMerkleHash calculates the root hash of events inserted into a
+// merkle trie with the hash of event as the key and encoded event as value
+func EventsListMerkleHash(el EventsList) (Identifier, error) {
+	var root Identifier
+	tree, err := merkle.NewTree(IdentifierLen)
+	if err != nil {
+		return root, err
+	}
 
 	for _, event := range el {
-		_, err := hasher.Write(event.Fingerprint())
+		// event fingerprint is the rlp encoding of the wrapperevent
+		// eventID is the standard sha3 hash of the event fingerprint
+		e := event
+		eventID := MakeID(e)
+		_, err = tree.Put(eventID[:], e.Fingerprint())
```

Comment:
MakeID uses fingerprinting, so we essentially call `e.Fingerprint()` twice here.

Reply:
👍  Changed it to reuse Fingerprint